utils/codepointtrie/src/codepointtrie.rs

```diff
@@ -302,6 +310,40 @@ impl<'trie, T: TrieValue> CodePointTrie<'trie, T> {
             .get(data_pos as usize)
             .unwrap_or(T::DATA_GET_ERROR_VALUE)
     }
+
+    /// Converts the CodePointTrie into one that returns another type of the same size.
+    ///
+    /// Borrowed data remains borrowed, and owned data remains owned.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new type `P` is a different size than the original type `T`.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use icu_codepointtrie::CodePointTrie;
+    ///
+    /// let cpt1: CodePointTrie<char> = todo!();
+    /// let cpt2: CodePointTrie<u32> = cpt1.convert_trie().expect("infallible");
+    /// ```
+    pub fn convert_trie<P>(
+        self,
+    ) -> Result<
+        CodePointTrie<'trie, P>,
+        <<P as zerovec::ule::AsULE>::ULE as zerovec::ule::ULE>::Error,
+    >
+    where
+        P: TrieValue,
+    {
+        assert_eq!(core::mem::size_of::<T>(), core::mem::size_of::<P>());
```

Comment:
thought: Not a fan of the panic, but I guess this is fine. We technically can use the `TrieValue` trait to enforce equality of underlying type (`P: TrieValue<Underlying = Self::Underlying>`). I slightly prefer the compile time safety but it's up to you, I know you already considered this.

Reply:
Yeah but how do you enforce the invariant on the associated type? Static assert? Make the trait unsafe?

Reply:
The invariant being that T::Underlying needs to be the same size as T

Reply:
You still use an `assert_eq!()` on sizeofs, however now there's a trait that means that users of CPT can't mess this up, only implementors of TrieValue (relatively rare).

Reply:
So the method still panics, but only if the implementer of TrieValue was wrong.  So we make the panic "rarer" but don't eliminate it, and we add a new associated type.

I would rather wait for the time when we can static_assert that the sizes are equal and keep the type parameter on the function.

Of course, this is subject to change depending in part on the resolution of #1183.

Reply:
> I would rather wait for the time when we can static_assert that the sizes are equal.

This is unlikely to ever be the case.

The new associated type is only a burden on implementors, and a rather straightforward one at that.

I think making it hard for ICU4X _users_ to cause a panic is valuable; even if it's possible for ICU4X implementors to incorrectly implement `TrieValue` and cause a panic (I don't see ICU4X users implementing that trait often)

Reply:
My only question was about this, too. We have the panicking.

I don't know how to evaluate the tradeoffs of what Manish is suggesting. If it's only a burden on implementors of TrieValue, and it doesn't happen too often (outside of a few primitive types and Unicode property enums/newtypes), to give us a reasonably good extra coverage of guarantees, even if not perfect, I guess that's preferable. Am I understanding this correctly?

Reply:
It sounds like part of @Manishearth's argument is an assertion that this trait, TrieValue, is primarily implemented by the ICU4X team.

Although this is true right now, *I don't see this as a long-term restriction.*  CodePointTrie is a utility crate that serves the very handy purpose of efficiently mapping code points to anything that is ZeroVec-compatible.  I definitely see it in the scope of responsibility longer-term that this trait TrieValue is implemented by users.

With an associated type, the place where the size requirement is needed and checked is far away from the place where the TrieValue impl can be found.  I fear that there could be TrieValue impls in the wild that have an incorrect `Underlying` type which is never noticed until someone files a bug.  This can happen easily by someone copy-pasting somone else's TrieType impl and not taking the time to read up on what all the associated types mean and what they're for.  Quite frankly, this could happen almost as easily in ICU4X code, due to a poor review, as it could happen outside ICU4X.

By putting the constraint directly on this method, all of the power is at the call site.  You aren't relying on your implementer to get it right.  And, importantly, if you get it wrong, you will find out right away, because the function unconditionally panics!  To be clear, I'm very anti-panic if the panic is *data-driven*, but I have no problem with them if they highlight *programmer error*.  This case is in the second category.

Another part of @Manishearth's argument seems to be that this is not good Rust style.  My counterpoint is to look at [std::vec::Vec](https://doc.rust-lang.org/std/vec/struct.Vec.html); it is filled with a lot of examples of methods that unconditionally panic if you give them an illegal argument.

Reply:
> Although this is true right now, _I don't see this as a long-term restriction._ CodePointTrie is a utility crate that serves the very handy purpose of efficiently mapping code points to anything that is ZeroVec-compatible. I definitely see it in the scope of responsibility longer-term that this trait TrieValue is implemented by users.

I still think that it will be _rare_ and as such moving the burden to one place (implementors) instead of having it be in two places (implementors and users) seems better. People aren't that careful about reading the docs of functions, people are more careful when implementing traits.

> Another part of @Manishearth's argument seems to be that this is not good Rust style. My counterpoint is to look at [std::vec::Vec](https://doc.rust-lang.org/std/vec/struct.Vec.html); it is filled with a lot of examples of methods that unconditionally panic if you give them an illegal argument.

It's not good Rust style to enforce something at runtime that can be enforced at compile time without undue burden. In the case of Vec, it's explicitly a runtime-length object so you can't enforce in bounds indexing; and `Option` based APIs would work (and they exist) except it's quite normal to want panicky indexing.

>  To be clear, I'm very anti-panic if the panic is _data-driven_, but I have no problem with them if they highlight _programmer error_. This case is in the second category.

I agree! And that's why I'm ultimately fine with this landing as-is, I'm just drawing a secondary distinction between "programmer error by library users" and "programmer error by library authors or power users".

Reply:
Okay, so as discussed earlier, I made the following change:

- Moved the assertion from `CodePointTrie::try_into_converted` to `ZeroVec::try_into_converted`

I then tried adding an associated type to `ULE`, which is already unsafe, so we could make this equivalence a requirement on the ULE trait safety.  However, I ran into problems because some of our types like `PairULE` have a size that needs to be computed in a const expression, which is not yet stable.  It may eventually be possible when something like `#![feature(generic_const_exprs)]` is fully implemented.

So I therefore suggest we stick with the panic for now.  At least we found a better, more correct home for it, with an offramp in the future.